indexing-service/src/main/java/org/apache/druid/indexing/common/Counters.java

```diff
@@ -19,54 +19,31 @@
 
 package org.apache.druid.indexing.common;
 
-import com.google.common.util.concurrent.AtomicDouble;
-
-import javax.annotation.Nullable;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicInteger;
-import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.BinaryOperator;
+import java.util.concurrent.atomic.AtomicLong;
 
-public class Counters
+public final class Counters
 {
-  private final ConcurrentMap<String, AtomicInteger> intCounters = new ConcurrentHashMap<>();
-  private final ConcurrentMap<String, AtomicDouble> doubleCounters = new ConcurrentHashMap<>();
-  private final ConcurrentMap<String, AtomicReference> objectCounters = new ConcurrentHashMap<>();
-
-  public int increment(String key, int val)
+  public static <K> int incrementAndGetInt(ConcurrentHashMap<K, AtomicInteger> counters, K key)
   {
-    return intCounters.computeIfAbsent(key, k -> new AtomicInteger()).addAndGet(val);
+    // get() before computeIfAbsent() is an optimization to avoid locking in computeIfAbsent() if not needed.
```

Comment:
Any idea why ConcurrentHashMap does not already employ an optimization like this?

Reply:
That's a throughput vs. scalability tradeoff, + lack of information. We are potentially doing two operations instead of one, and avoid locking in some cases instead.

At some sites where `computeIfAbsent()` is actually expected to find the key absent and compute the value most of the time, the `get()` guard just makes things worse.

There is also an area where it's hard for me to say what approach is better, is that when the map is big and `computeIfAbsent()` constitutes significant part of the app's CPU consumption (the bigger the map and the hotter `computeIfAbsent()` call is, the more likely that it's better to not guard `computeIfAbsent()` with `get()`). I think it's never nearly the case on Druid nodes that `computeIfAbsent()` is hot, but I could be wrong.

Reply:
From the ConcurrentHashMap's part, it would be useful if `computeIfAbsentMoreScalableButMaybeDoingExtraWork()` existed, where they don't recompute hash bucket twice and just walk the collision chain twice. But it's easy to imagine why such method doesn't exist.